docs/dev/design/better-doc-urls-handling.rst

```diff
@@ -0,0 +1,417 @@
+Better handling of docs URLs
+============================
+
+``proxito`` is the component of our code base in charge of serving documentation
+to users and handling any other URLs from the user documentation domain.
+
+The current implementation has some problems that are discussed in this document,
+and an alternative implementation is proposed to solve those problems.
+
+Current implementation
+----------------------
+
+The current implementation is based on Django URLs
+trying to match a pattern that looks like a single project, a versioned project,
+or a subproject, this means that a couple or URLs are *reserved*,
+and won't resolve to the correct file if it exists
+(https://github.com/readthedocs/readthedocs.org/issues/8399, https://github.com/readthedocs/readthedocs.org/issues/2292),
+this usually happens with single version projects.
+
+And to support custom URLs we are hacking into Django's urlconf
+to override it at runtime,
+this doesn't allow us to implement custom URLs for subprojects easily
+(https://github.com/readthedocs/readthedocs.org/pull/8327).
+
+Alternative implementation
+--------------------------
+
+Instead of trying to map a URL to a view,
+we first analyze the canonical project (given from the subdomain),
+and based on that we map each part of the URL (parts are the result of splitting the URL on ``/``)
+to the *current* project.
```

Comment:
This approach will still require you to test with several regexes, if a project has subprojects it doesn't necessarily mean that it will always serve a subproject, this will also have the same problem of reserved paths, like a single version project with subprojects won't be able to have a `projects` path, since the regex will match that path as a subproject path instead of a `projects` directory.

Also, we already have all the information, we don't want to keep guessing the current project (with this proposal we are still guessing, but just the requested file, the current project is already inferred from the canonical project and the URL).